src/System.IO.Compression/src/System/IO/Compression/DeflateStream.cs

```diff
@@ -0,0 +1,644 @@
+                // the object get disposed because eventually ZLibNative.ReleaseHandle will get called during 
+                // the dispose operation and although it frees the stream but it return error code because the 
+                // stream state was still marked as in use. The symptoms of this problem will not be seen except 
+                // if running any diagnostic tools which check for disposing safe handle objects
+                bool finished;
+                do
+                {
+                    int compressedBytes;
+                    finished = deflater.Finish(buffer, out compressedBytes);
+                } while (!finished);
+            }
+
+            // Write format footer:
+            if (formatWriter != null && wroteHeader)
+            {
+                byte[] b = formatWriter.GetFooter();
+                _stream.Write(b, 0, b.Length);
+            }
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            try
+            {
+                PurgeBuffers(disposing);
+            }
+            finally
+            {
+                // Close the underlying stream even if PurgeBuffers threw.
+                // Stream.Close() may throw here (may or may not be due to the same error).
+                // In this case, we still need to clean up internal resources, hence the inner finally blocks.
+                try
+                {
+                    if (disposing && !_leaveOpen && _stream != null)
+                        _stream.Dispose();
+                }
+                finally
+                {
+                    _stream = null;
+
+                    try
+                    {
+                        if (deflater != null)
+                            deflater.Dispose();
+                    }
+                    finally
+                    {
+                        deflater = null;
+                        base.Dispose(disposing);
+                    }
+                }  // finally
+            }  // finally
+        }  // Dispose
+
+        public override Task WriteAsync(Byte[] array, int offset, int count, CancellationToken cancellationToken)
+        {
+            EnsureCompressionMode();
+
+            // We use this checking order for compat to earlier versions:
+            if (asyncOperations != 0)
+                throw new InvalidOperationException(SR.InvalidBeginCall);
+
+            ValidateParameters(array, offset, count);
+            EnsureNotDisposed();
+
+            if (cancellationToken.IsCancellationRequested)
+                return TaskHelpers.FromCancellation<int>(cancellationToken);
+
+            Interlocked.Increment(ref asyncOperations);
+
+            try
+            {
+                return base.WriteAsync(array, offset, count, cancellationToken).ContinueWith(
+                        (t) => Interlocked.Decrement(ref asyncOperations),
```

Comment:
This looks like a real bug.  Any exceptions or cancellation information from the Task returned from the base WriteAsync is getting eaten.